File: AvnCanvasHelper/CanvasHelper.razor.js
/*
 * Adapted from Scott Harden's EXCELLENT blog post, 
 * "Draw Animated Graphics in the Browser with Blazor WebAssembly"
 * https://swharden.com/blog/2021-01-07-blazor-canvas-animated-graphics/
 */

/*This is called from the Blazor component's Initialize method*/
export function initRenderJS(instance) {
    // instance is the Blazor component dotnet reference
    window.theInstance = instance;

    // tell the window we want to handle the resize event
    window.addEventListener("resize", resizeCanvasToFitWindow);

    // ... and the mouse events
    window.addEventListener("mousedown", mouseDown);
    window.addEventListener("mouseup", mouseUp);
    window.addEventListener("mousemove", mouseMove);

    // Call resize now
    resizeCanvasToFitWindow();
    // request an animation frame, telling window to call renderJS
    // https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame
    window.requestAnimationFrame(renderJS);
}

/*This is called whenever we have requested an animation frame*/
function renderJS(timeStamp) {
    // Call the blazor component's [JSInvokable] RenderInBlazor method
    theInstance.invokeMethodAsync('RenderInBlazor', timeStamp);
    // request another animation frame
    window.requestAnimationFrame(renderJS);
}

/*This is called whenever the browser (and therefore the canvas) is resized*/
function resizeCanvasToFitWindow() {
    // canvasHolder is the ID of the div that wraps the renderfragment 
    // content(Canvas) in the blazor app
    var holder = document.getElementById('canvasHolder');
    // find the canvas within the renderfragment
    var canvas = holder.querySelector('canvas');
    if (canvas) {
        // resize the canvas
        canvas.width = window.innerWidth;
        canvas.height = window.innerHeight;
        // Call the blazor component's [JSInvokable] ResizeInBlazor method
        theInstance.invokeMethodAsync('ResizeInBlazor', canvas.width, canvas.height);
    }
}

//Handle the window.mousedown event
function mouseDown(e) {
    var args = canvasMouseMoveArgs(e);
    theInstance.invokeMethodAsync('OnMouseDown', args);
}

//Handle the window.mouseup event
function mouseUp(e) {
    var args = canvasMouseMoveArgs(e);
    theInstance.invokeMethodAsync('OnMouseUp', args);
}

//Handle the window.mousemove event
function mouseMove(e) {
    var args = canvasMouseMoveArgs(e);
    theInstance.invokeMethodAsync('OnMouseMove', args);
}

// Extend the CanvasMouseArgs.cs class (and this) as necessary
function canvasMouseMoveArgs(e) {
    return {
        ScreenX: e.screenX,
        ScreenY: e.screenY,
        ClientX: e.clientX,
        ClientY: e.clientY,
        MovementX: e.movementX,
        MovementY: e.movementY,
        OffsetX: e.offsetX,
        OffsetY: e.offsetY,
        AltKey: e.altKey,
        CtrlKey: e.ctrlKey,
        Button: e.button,
        Buttons: e.button,
        Bubbles: e.bubbles
    };
}

File: AvnCanvasHelper/CanvasHelper.razor.cs
/*
 * Adapted from Scott Harden's EXCELLENT blog post, 
 * "Draw Animated Graphics in the Browser with Blazor WebAssembly"
 * https://swharden.com/blog/2021-01-07-blazor-canvas-animated-graphics/
 */

using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;
using System.Drawing;

namespace AvnCanvasHelper
{
    /// <summary>
    /// CanvasHelper component gives you render and resize callbacks for Canvas animation
    /// </summary>
    public partial class CanvasHelper : ComponentBase, IAsyncDisposable
    {
        /// <summary>
        /// JS Interop module used to call JavaScript
        /// </summary>
        private Lazy<Task<IJSObjectReference>> _moduleTask;
        
        /// <summary>
        /// Used to calculate the frames per second
        /// </summary>
        private DateTime _lastRender;

        /// <summary>
        /// JS Runtime
        /// </summary>
        [Inject]
        protected IJSRuntime _jsRuntime { get; set; }

        /// <summary>
        /// Event called when the browser (and therefore the canvas) is resized
        /// </summary>
        [Parameter]
        public EventCallback<Size> CanvasResized { get; set; }

        /// <summary>
        /// Event called every time a frame can be redrawn
        /// </summary>
        [Parameter]
        public EventCallback<double> RenderFrame { get; set; }

        /// <summary>
        /// Event called on mouse down
        /// </summary>
        [Parameter]
        public EventCallback<CanvasMouseArgs> MouseDown { get; set; }

        /// <summary>
        /// Event called on mouse up
        /// </summary>
        [Parameter]
        public EventCallback<CanvasMouseArgs> MouseUp { get; set; }

        /// <summary>
        /// Event called on mouse move
        /// </summary>
        [Parameter]
        public EventCallback<CanvasMouseArgs> MouseMove { get; set; }


        /// <summary>
        /// Call this in your Blazor app's OnAfterRenderAsync method when firstRender is true
        /// </summary>
        /// <returns></returns>
        public async Task Initialize()
        {
            // We need to specify the .js file path relative to this code
            _moduleTask = new(() => _jsRuntime.InvokeAsync<IJSObjectReference>(
                "import", "./_content/AvnCanvasHelper/CanvasHelper.razor.js").AsTask());

            // Load the module
            var module = await _moduleTask.Value;

            // Initialize
            await module.InvokeVoidAsync("initRenderJS", DotNetObjectReference.Create(this));
            
            // Dispose the module
            await module.DisposeAsync();
        }

        /// <summary>
        /// Handle the JavaScript event called when the browser/canvas is resized
        /// </summary>
        /// <param name="width"></param>
        /// <param name="height"></param>
        /// <returns></returns>
        [JSInvokable]
        public async Task ResizeInBlazor(int width, int height)
        {
            var size = new Size(width, height);
            // Raise the CanvasResized event to the Blazor app
            await CanvasResized.InvokeAsync(size);
        }

        /// <summary>
        /// Handle the JavaScript event when a frame is ready to render
        /// </summary>
        /// <param name="timeStamp"></param>
        /// <returns></returns>
        [JSInvokable]
        public async ValueTask RenderInBlazor(float timeStamp)
        {
            // calculate frames per second
            double fps = 1.0 / (DateTime.Now - _lastRender).TotalSeconds;
            
            _lastRender = DateTime.Now; // update for the next time 

            // raise the RenderFrame event to the blazor app
            await RenderFrame.InvokeAsync(fps);
        }

        /// <summary>
        /// Handle the JavaScript window.mousedown event
        /// </summary>
        /// <param name="args"></param>
        /// <returns></returns>
        [JSInvokable]
        public async Task OnMouseDown(CanvasMouseArgs args)
        {
            await MouseDown.InvokeAsync(args);
        }

        /// <summary>
        /// Handle the JavaScript window.mouseup event
        /// </summary>
        /// <param name="args"></param>
        /// <returns></returns>
        [JSInvokable]
        public async Task OnMouseUp(CanvasMouseArgs args)
        {
            await MouseUp.InvokeAsync(args);
        }

        /// <summary>
        /// Handle the JavaScript window.mousemove event
        /// </summary>
        /// <param name="args"></param>
        /// <returns></returns>
        [JSInvokable]
        public async Task OnMouseMove(CanvasMouseArgs args)
        {
            await MouseMove.InvokeAsync(args);
        }

        /// <summary>
        /// Dispose of our module resource
        /// </summary>
        /// <returns></returns>
        public async ValueTask DisposeAsync()
        {
            if (_moduleTask != null && _moduleTask.IsValueCreated)
            {
                var module = await _moduleTask.Value;
                await module.DisposeAsync();
            }
        }
    }
}
